### ISSUE TITLE
Trouble replicating paper's result on GiraffeZebraID

### ISSUE THREAD

**Issue by VincentCoulombe** [WildlifeDatasets/wildlife-tools#23]
Hi, 

I would like to preface my issue with a congratulation for the nice work you've been doing. It will make my life so much easier on experimenting with re-identification datasets once I figure out how to run the pipelines properly.

That said, I must be doing something wrong, as I cannot replicate (or get close to) the paper's reported level of performance on GiraffeZebraID (66%). In fact: The following prints an accuracy of 5%:  

"""
from wildlife_datasets.datasets import GiraffeZebraID
from wildlife_tools.data import ImageDataset
import torchvision.transforms as T
import timm
from wildlife_tools.features import DeepFeatures
from wildlife_tools.similarity import CosineSimilarity
import numpy as np
from wildlife_tools.inference import KnnClassifier

metadata = GiraffeZebraID("../wildlife-datasets/data/GiraffeZebraID")
transform = T.Compose(
    [T.Resize([224, 224]), T.ToTensor(), T.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))]
)
dataset = ImageDataset(metadata.df, metadata.root, transform=transform)

dataset_database = ImageDataset(metadata.df.iloc[100:, :], metadata.root, transform=transform)
dataset_query = ImageDataset(metadata.df.iloc[:100, :], metadata.root, transform=transform)

name = "hf-hub:BVRA/MegaDescriptor-T-224"
extractor = DeepFeatures(timm.create_model(name, num_classes=0, pretrained=True))
query, database = extractor(dataset_query), extractor(dataset_database)

similarity_function = CosineSimilarity()
similarity = similarity_function(query, database)

classifier = KnnClassifier(k=1, database_labels=dataset_database.labels_string)
predictions = classifier(similarity)
accuracy = np.mean(dataset_query.labels_string == predictions)
print(accuracy)
"""

Obviously, I did something wrong here. I am simply wondering what. Your input would save me a lot of time. 

Thank you,

Vincent

**Comment by picekl** [comment]
Dear @VincentCoulombe,

I think the main problem is in the way of selecting data for the database and query. You simply take the first 100 rows and use them for the database, which most likely will result in "most individuals" not being present in the database.
```
dataset_database = ImageDataset(metadata.df.iloc[100:, :], metadata.root, transform=transform)
dataset_query = ImageDataset(metadata.df.iloc[:100, :], metadata.root, transform=transform)
```

The database should include all possible identities. Therefore, your selection should be a little bit smarter and consider different identities. You can read more about it in [this paper](https://arxiv.org/pdf/2406.09211).

Best,
Lukas

**Comment by VincentCoulombe** [comment]
Thank you, the publication was indeed very insightful. One of the biggest eye-openers, other than the answer to my question, was the gap favoring the CE loss over the ArcFace loss (Figure 4). As I am very interested in leveraging your framework for my work, I would like to know if your CE implementation (SoftmaxLoss) correspond to the loss function used in the publication (IE without any mining and with a linear layer for classification)?

**Comment by sadda** [comment]
Hi @VincentCoulombe,

I would like to also add that we used

```
name = "hf-hub:BVRA/MegaDescriptor-L-384"
```

instead of `T-224` as you mentioned. You can check the [inference scripts](https://github.com/WildlifeDatasets/wildlife-tools/blob/main/baselines/inference/MegaDescriptor-L-384.ipynb) to get a bit more insight.

I am closing the issue but feel free to ask further questions :)

Best,

Lukas

**Comment by VincentCoulombe** [comment]
Thank you for your swift responses :) I was originally planning on using the good old TripletLoss, but I've been reading that a CE loss in the context of ReID could lead to better results (I think I first read it in the FairMOT paper, though I am not certain). And yours also suggest the same. That said, I am eager to experiment with such a loss function, but I am lacking directions on how to apply it to the ReID task. That said, I saw in your repository that you have an implementation named "SoftmaxLoss" that pairs the CE loss with a linear layer. As such, I was wondering if this is the implementation that led to good results in your publication, and what is your intuition on how it could outperform ArcFace and Triplet. Thank you again for your contributions to the field :)

**Comment by sadda** [comment]
Hi @VincentCoulombe , if you are interested in the exact architecture of MegaDescriptor, you can check our [training scripts](https://github.com/WildlifeDatasets/wildlife-tools/blob/main/baselines/training/MegaDescriptor-L-384.ipynb), from whose the architecture can be deduced.

Best, Lukas
